File: frontend/lib/features/task/domain/services/due_grouping.dart
import '../entities/task.dart';

/// Ordered buckets for the smart due-date grouping. Completed tasks
/// always collapse into their own final section, whatever their date:
/// urgency grouping is for work still to do.
enum DueGroup { overdue, today, tomorrow, thisWeek, later, noDate, completed }

const dueGroupLabels = {
  DueGroup.overdue: 'In ritardo',
  DueGroup.today: 'Oggi',
  DueGroup.tomorrow: 'Domani',
  DueGroup.thisWeek: 'Questa settimana',
  DueGroup.later: 'Più avanti',
  DueGroup.noDate: 'Senza scadenza',
  DueGroup.completed: 'Completati',
};

final class DueSection {
  const DueSection(this.group, this.tasks);

  final DueGroup group;
  final List<Task> tasks;

  String get label => dueGroupLabels[group]!;
}

/// The single overdue rule, shared by grouping, card highlighting and
/// (conceptually) the backend analytics: past due AND not done.
bool isOverdue(Task task, DateTime now) =>
    task.status != TaskStatus.done &&
    task.dueDate != null &&
    task.dueDate!.toLocal().isBefore(now);

/// Pure function: buckets tasks into ordered sections using LOCAL
/// calendar-day boundaries. Empty sections are omitted.
List<DueSection> groupByDue(List<Task> tasks, DateTime now) {
  final today = DateTime(now.year, now.month, now.day);

  DueGroup classify(Task task) {
    if (task.status == TaskStatus.done) {
      return DueGroup.completed;
    }
    final due = task.dueDate?.toLocal();
    if (due == null) {
      return DueGroup.noDate;
    }
    if (due.isBefore(now)) {
      return DueGroup.overdue;
    }
    final dueDay = DateTime(due.year, due.month, due.day);
    final daysAhead = dueDay.difference(today).inDays;
    return switch (daysAhead) {
      0 => DueGroup.today,
      1 => DueGroup.tomorrow,
      >= 2 && <= 6 => DueGroup.thisWeek,
      _ => DueGroup.later,
    };
  }

  final buckets = <DueGroup, List<Task>>{};
  for (final task in tasks) {
    buckets.putIfAbsent(classify(task), () => []).add(task);
  }

  return [
    for (final group in DueGroup.values)
      if (buckets.containsKey(group)) DueSection(group, buckets[group]!),
  ];
}

File: webapp/src/features/tasks/due-grouping.ts
import { differenceInCalendarDays } from 'date-fns'
import type { Task } from '@/api/types'

/**
 * Port di `due_grouping.dart`. Funzione pura con il tempo iniettato: è
 * l'unico modo per testare "oggi" senza aspettare domani.
 *
 * I completati collassano sempre in una sezione a parte, qualunque sia la
 * loro scadenza: il raggruppamento per urgenza riguarda ciò che resta da fare.
 */
export const dueGroups = [
  'overdue',
  'today',
  'tomorrow',
  'thisWeek',
  'later',
  'noDate',
  'completed',
] as const

export type DueGroup = (typeof dueGroups)[number]

export const dueGroupLabels: Record<DueGroup, string> = {
  overdue: 'In ritardo',
  today: 'Oggi',
  tomorrow: 'Domani',
  thisWeek: 'Questa settimana',
  later: 'Più avanti',
  noDate: 'Senza scadenza',
  completed: 'Completati',
}

export interface DueSection {
  group: DueGroup
  label: string
  tasks: Task[]
}

/**
 * La regola unica di "in ritardo", condivisa da raggruppamento, evidenziazione
 * nella card e — concettualmente — dalle analytics del backend: scaduto e non
 * completato.
 */
export function isOverdue(task: Task, now: Date): boolean {
  return task.status !== 'DONE' && task.dueDate !== undefined && new Date(task.dueDate) < now
}

function classify(task: Task, now: Date): DueGroup {
  if (task.status === 'DONE') {
    return 'completed'
  }
  if (task.dueDate === undefined) {
    return 'noDate'
  }

  const due = new Date(task.dueDate)
  if (due < now) {
    return 'overdue'
  }

  // Differenza in giorni di *calendario* locale: le 23:00 di oggi e le 01:00
  // di domani distano due ore ma un giorno, ed è il giorno che conta.
  const daysAhead = differenceInCalendarDays(due, now)
  if (daysAhead === 0) {
    return 'today'
  }
  if (daysAhead === 1) {
    return 'tomorrow'
  }
  return daysAhead <= 6 ? 'thisWeek' : 'later'
}

/** Raggruppa in sezioni ordinate per urgenza; le sezioni vuote non compaiono. */
export function groupByDue(tasks: Task[], now: Date): DueSection[] {
  const buckets = new Map<DueGroup, Task[]>()

  for (const task of tasks) {
    const group = classify(task, now)
    const bucket = buckets.get(group)
    if (bucket) {
      bucket.push(task)
    } else {
      buckets.set(group, [task])
    }
  }

  return dueGroups
    .filter((group) => buckets.has(group))
    .map((group) => ({ group, label: dueGroupLabels[group], tasks: buckets.get(group)! }))
}
